pkg/calculator/experimentcalc/experiment_calculator.go

```diff
@@ -0,0 +1,650 @@
+			gr.GoalId = experiment.GoalId
+			e.appendVariationResult(ctx, timestamp, goalResult, gr.VariationResults)
+		}
+		experimentResult.GoalResults = append(experimentResult.GoalResults, goalResult)
+	}
+
+	return experimentResult
+}
+
+func (e ExperimentCalculator) listEnvironments(
+	ctx context.Context,
+) ([]*environment.Environment, error) {
+	listEnvironmentsRequest := environment.ListEnvironmentsRequest{
+		PageSize: 0,
+		Cursor:   "",
+	}
+	resp, err := e.environmentClient.ListEnvironments(ctx, &listEnvironmentsRequest)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Environments, err
+}
+
+func (e ExperimentCalculator) listExperiments(
+	ctx context.Context,
+	namespace string,
+) ([]*experiment.Experiment, error) {
+	req := &experiment.ListExperimentsRequest{
+		From:                 time.Now().Add(-2 * 24 * time.Hour).Unix(),
+		PageSize:             0,
+		Cursor:               "",
+		EnvironmentNamespace: namespace,
+		Statuses: []experiment.Experiment_Status{
+			experiment.Experiment_WAITING,
+			experiment.Experiment_RUNNING,
+		},
+	}
+	resp, err := e.experimentClient.ListExperiments(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Experiments, nil
+}
+
+func (e ExperimentCalculator) getEvaluationCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentEvaluationCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	evaluationCount, err := e.eventCounterClient.GetExperimentEvaluationCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range evaluationCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) getGoalCount(
+	ctx context.Context,
+	req *eventcounter.GetExperimentGoalCountRequest,
+) (map[string]*eventcounter.VariationCount, error) {
+	variationCounts := make(map[string]*eventcounter.VariationCount)
+	goalCount, err := e.eventCounterClient.GetExperimentGoalCount(ctx, req)
+	if err != nil {
+		return variationCounts, err
+	}
+	for _, variationCount := range goalCount.VariationCounts {
+		variationCounts[variationCount.VariationId] = variationCount
+	}
+	return variationCounts, nil
+}
+
+func (e ExperimentCalculator) calcGoalResult(
+	ctx context.Context,
+	evalVariationCounts,
+	goalVariationCounts map[string]*eventcounter.VariationCount,
+	baseVid string,
+) *eventcounter.GoalResult {
+	goalResult := &eventcounter.GoalResult{}
+	length := len(goalVariationCounts)
+	vids := make([]string, 0, length)
+	goalUc, evalUc := make([]int64, 0, length), make([]int64, 0, length)
+	vrs := make(map[string]*eventcounter.VariationResult, length)
+	valueMeans, valueVars := make([]float64, 0, length), make([]float64, 0, length)
+	baselineIdx, loopIdx := 0, 0
+
+	for vid, goalVariationCount := range goalVariationCounts {
+		if _, ok := evalVariationCounts[vid]; !ok {
```

Comment:
I think, at this point, the variation count should be on the map. If not, it should be treated as an error.
What do you think?

Either way, we should add a metric here so we can monitor those cases.

E.g.

```go
evalVariationCountNotFound
```

Reply:
I think so too. But I was not sure how to optimize this when developing and just keep this logic the same as Python code, so we can discuss about this here.

Anyway, like you said, we should ad a metric here to let us know when this case happens.

Reply:
Reference.
https://github.com/bucketeer-io/bucketeer/pull/395#discussion_r1243356931